keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1107 @@
+  apiservers are rolledback.
+
+But a node-by-node upgrade or rollback does not achieve this today.
+
+- For upgrade, there is about a 25% chance of a new version of the controller
+  running while old versions of the apiserver are active, resulting in a skew
+  violation. (Consider case where the 2nd node upgraded has the lease)
+- For rollback, it is almost a certainty that skew will be violated.
+
+There is also the possiblity that the lease will be lost by a leader during an
+upgrade or rollback resulting in the version of the controller flip-flopping
+between old and new.
+
+### Goals
+
+During HA upgrades/rollbacks/downgrades, 
+
+Leader elected components:
+
+- Change versions at predictable times
+- Do not violate version skew, even during node-by-node rollbacks
+
+The control plane:
+
+- Can safely canary components and nodes at the new version for an extended
+  period of time, or to pause an upgrade at any step durning an upgrade. This
+  enhancement, combined with
+  [UVIP](../4020-unknown-version-interoperability-proxy) helps achieve this.
+  
+
+### Non-Goals
+
+- Change the default leader election for components.
+
+## Proposal
+
+- Offer an opt-in leader election mechanism to:
+  - Elect the candidate with the oldest version available.
+  - Provide a way to preempt the current leader.
+  - Reuse the existing lease mechanism as much as possible.
+
+### Component Identity Leases
+
+Components will create identity leases similar to those used by apiserver identity, e.g.:
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  annotations:
+    coordination.k8s.io/binary-version: "1.29"
+    coordination.k8s.io/can-lead-leases: kube-system/sample-controller
+    coordination.k8s.io/compatibility-version: "1.29"
+  name: sample-controller-0001A
+  namespace: kube-system
+spec:
+  holderIdentity: sample-controller-0001A
+  leaseDurationSeconds: 10
+  renewTime: "2023-12-05T02:33:08.685777Z"
+```
+
+A component identity lease announces candidacy for leadership by including
+`coordination.k8s.io/can-lead-leases` it is identity lease. If the lease
+expires, the component is considered unavailable for leader election purposes.
+
+### Coordinated Election Controller
+
+A Coordinated Election Controller will reconcile component Leader Leases
+(primary resource) and Identity Leases (secondary resource, changes trigger
+reconciliation of related leader leases).
+
+Coordinated Election Controller reconciliation loop:
+
+- If no leader lease exists for a components:
+  - Elect leader from candidates by preparing a freshly renewed lease with:
+    - `spec.holderIdentity` set to the identity of the elected leader
+    - `coordination.k8s.io/elected-by: leader-election-controller` (to make
+      lease types easy to disambiguate)
+- If there is a better candidate than current leader:
+  - Sets `coordination.k8s.io/end-of-term: true` on the leader lease, signaling
```

Comment:
this requires all older client-go to have migrated to the version which has the impl. Is that the right way to think? The example works with the lease object, but i think given all candidates have to express their availability and then a central authority picks the right candidate, there is a need to create another object type that contains all candidates.